sklearn/impute/_knn.py

```diff
@@ -115,6 +117,14 @@ class KNNImputer(_BaseImputer):
            [8. , 8. , 7. ]])
     """
 
+    _parameter_constraints = {
+        **_BaseImputer._parameter_constraints,
+        "n_neighbors": [Interval(Integral, 1, None, closed="left")],
+        "weights": [StrOptions({"uniform", "distance"}), callable],
```

Comment:
Although not documented, `None` is currently a valid option for `weights`, since eventually we set `weight_matrix = _get_weights(donors_dist, self.weights)` and in `_get_weights` it just checks `weights in (None, "uniform")`. This is also the case in `KNeighborsClassifier`. I suspect we probably want this to remain undocumented? In which case we can use `..utils._param_validation.Hidden`:

```suggestion
        "weights": [StrOptions({"uniform", "distance"}), callable, Hidden(None)],
```